evm/p2p/lightchain.py

```diff
@@ -59,25 +60,43 @@ class LightChain(Chain):
     def __init__(self, chaindb: BaseChainDB) -> None:
         super(LightChain, self).__init__(chaindb)
         self.peer_pool = self.peer_pool_class(
-            LESPeer, chaindb, self.network_id, self.privkey, self.msg_handler)
+            LESPeer, chaindb, self.network_id, self.privkey, self.new_peer_callback)
         self._announcement_queue = asyncio.Queue()  # type: asyncio.Queue[Tuple[LESPeer, les.HeadInfo]]  # noqa: E501
         self._last_processed_announcements = {}  # type: Dict[LESPeer, les.HeadInfo]
-        self._latest_head_info = {}  # type: Dict[LESPeer, les.HeadInfo]
         self._should_stop = asyncio.Event()
         self._finished = asyncio.Event()
 
-    def msg_handler(self, peer: BasePeer, cmd: protocol.Command,
-                    announcement: protocol._DecodedMsgType) -> None:
-        """The callback passed to BasePeer, called for every incoming message."""
-        peer = cast(LESPeer, peer)
-        if isinstance(cmd, (les.Announce, les.Status)):
-            head_info = cmd.as_head_info(announcement)
-            self._latest_head_info[peer] = head_info
-            self._announcement_queue.put_nowait((peer, head_info))
+    def new_peer_callback(self, peer: BasePeer) -> None:
+        asyncio.ensure_future(self.handle_peer(cast(LESPeer, peer)))
+
+    async def handle_peer(self, peer: LESPeer) -> None:
+        """Handle the lifetime of the given peer.
+
+        Returns when the peer is finished or when the LightChain is asked to stop.
+        """
+        try:
+            await peer.do_sub_proto_handshake()
+        except HandshakeFailure as e:
+            self.logger.info(str(e))
+            return
+        await self._handle_peer(peer)
+        self.logger.info("Peer %s finished", peer)
+
+    async def _handle_peer(self, peer: LESPeer) -> None:
```

Comment:
This function is confusing me.  I can't tell what it's supposed to be doing or why it's doing what it is doing.

It seems to only handle the `les.Announce` command, silently ignoring all others and discarding them indefinitely untill `PeerFinished` kills the loop ad disconnects the peer.  The silently ignoring other commands is the confusing part for me.

Reply:
That's because (apart from `Announce`) we implement only the LES commands for retrieving data, and since those will always come in response to requests we make they are consumed by `LESPeer.handle_sub_proto_msg()`. I'll add a comment to that effect, and change the code to raise an `UnexpectedMessage` if we get anything other than an `Announce` as that would indicate a bug in the peer's handling of sub-proto msgs